simperby/src/node.rs

```diff
@@ -0,0 +1,195 @@
+                run_network_task(network, state_).await;
+            }),
+            genesis_info,
+        })
+    }
+}
+
+impl Drop for Node {
+    fn drop(&mut self) {
+        self.network_task.abort();
+    }
+}
+
+#[async_trait]
+impl SimperbyApi for Node {
+    fn get_genesis_info(&self) -> &GenesisInfo {
+        &self.genesis_info
+    }
+
+    async fn get_height(&self) -> u64 {
+        self.state.read().await.last_header.height
+    }
+
+    async fn get_block(&self, height: BlockHeight) -> Result<Block, SimperbyError> {
+        let state = self.state.read().await;
+        if state.last_header.height < height {
+            return Err(SimperbyError::InvalidOperation(format!(
+                "height {} is greater than the current height {}",
+                height, state.last_header.height
+            )));
+        }
+
+        state
+            .history_storage
+            .get_block(height)
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn check_block(&self, _block: Block, _height: BlockHeight) -> Result<(), SimperbyError> {
+        unimplemented!()
+    }
+
+    async fn read_state(&self, key: String, height: u64) -> Result<Vec<u8>, SimperbyError> {
+        let state = self.state.read().await;
+        assert_eq!(
+            height, state.last_header.height,
+            "currently state query is supported only for the last block"
+        );
+        state
+            .state_storage
+            .get(Hash256::hash(key))
+            .await
+            .map_err(SimperbyError::StorageError)
+    }
+
+    async fn get_consensus_vote_options(&self) -> Result<Vec<ConsensusVoteItem>, SimperbyError> {
+        unimplemented!()
+    }
+
+    async fn get_consensus_status(&self) -> () {
+        unimplemented!()
+    }
+
+    async fn get_network_status(&self) -> () {
+        unimplemented!()
+    }
+
+    /// Gets the `number`-last logs of attempts to execute `SimperbyOperation`s.
+    async fn get_operation_log(&self, _number: usize) -> Vec<SimperbyOperationLog> {
+        unimplemented!()
+    }
+
+    async fn propose_block(
+        &self,
+        block: Block,
+        signature: TypedSignature<Hash256>,
+    ) -> Result<(), SimperbyError> {
+        self.state
+            .write()
+            .await
+            .submit_block_proposal(block, signature)
+    }
+
+    async fn submit_consensus_vote(
+        &self,
+        _hash: Hash256,
+        _signature: TypedSignature<Hash256>,
+    ) -> Result<(), SimperbyError> {
+        unimplemented!()
+    }
+}
+
+async fn run_network_task(
+    _network: Box<dyn AuthorizedNetwork>,
+    _node_state: Arc<RwLock<NodeState>>,
+) {
+    unimplemented!()
+}
+
+/// The node state machine.
+///
+/// Both `SimperbyApi` and `run_network_task()` can concurrently access to this state
+/// and it will be synchronized by `RwLock`.
+struct NodeState {
+    history_storage: HistoryStorage,
+    state_storage: Box<dyn KVStorage>,
```

Comment:
why is this `Box<dyn KVStorage>`, while history_storage is `HistoryStorage`?
Does something else make use of this storage?

Reply:
Unlike `HistoryStorage`, it is not obvious yet that we need an additional layer that wraps `KVStorage`, for the case of state storage. Let's keep that and revisit later.

Reply:
Then how about a TODO comment or an issue?

Reply:
fixed.